rfcs/convergence/slot-null-rendering.md

```diff
@@ -0,0 +1,261 @@
+       * if this was a hook than it would need
+       * to be declared before the conditional
+       **/
+    };
+  }
+
+  useAccordionHeaderStyles(state);
+  return renderAccordionHeader(state);
+});
+```
+
+In cases where we simply want to ensure that the slot should be rendered even when `shorthand` is `undefined` than `{optional: false}` should be provided
+
+```ts
+export const useState = (props, ref) => {
+  return {
+    ...props,
+    slot1: resolveShorthand(props.slot1), // undefined
+    slot2: resolveShorthand(props.slot2, {}, { optional: false }), // {}
+  };
+};
+```
+
+### Option 2: Symbol for null rendering
+
+This solves this problem by verifying if the shorthand has a special symbol or not.
+By verifying this symbol we can opt for null rendering without compromising the cases which native slots don't have children.
+This is very similar to Option 1 approach but without the downside of having undefined as part of `ObjectShorthandProps` interface.
+
+- `getSlots` should stop verifying null rendering by the existence of the property `children`, instead depending on the existence of the symbol
+
+```tsx
+// react-utilities/src/compose/getSlots.ts
+function getSlots(state, slotNames) {
+  // ...
+  if (typedState[name][nullRenderSymbol]) {
+    slots[name] = nullRender;
+    continue;
+  }
+  // ...
+}
+```
+
+- Since some slots may have default props, verification must be done on shorthand declaration to ensure that those slots will be null rendered even presenting default props
+
+```diff
+export function resolveShorthand(
+	value,
+	defaultProps,
++   { optional = true },
++) {
++  if (value === null || (value === undefined && optional)) {
++    return { [nullRenderSymbol]: true };
++  }
+```
+
+The difference with this approach is that the shorthand signature will not carry with it an `undefined` value, making overrides easier:
+
+```ts
+export const CustomAccordionHeader = React.forwardRef<HTMLElement, AccordionHeaderProps>((props, ref) => {
+  const state = useAccordionHeader(props, ref);
+  state.icon.onClick = () => {
+    /* ... */
+  };
+  useAccordionHeaderStyles(state);
+  return renderAccordionHeader(state);
+});
+```
+
+In cases where we simply want to ensure that the slot should be rendered even when `shorthand` is `undefined` than `{optional: false}` should be provided
+
+```ts
+export const useState = (props, ref) => {
+  return {
+    ...props,
+    slot1: resolveShorthand(props.slot1), // undefined
+    slot2: resolveShorthand(props.slot2, {}, { optional: false }), // {}
+  };
+};
+```
+
+## Pros and Cons
+
+### Pros
+
+1. This takes the responsibility of deciding if a slot might be null rendered out of inner implementation of `getSlot` into the developers hand, by providing a optional way of declaring it
+
+### Cons
+
+1. This takes the responsibility of deciding if a slot might be null rendered out of inner implementation of `getSlot` into the developers hand, by providing a optional way of declaring it
```

Comment:
This is just a copy of the pro above, is that intended?

Reply:
Yes, it's intentional.  Making shorthand logic more explicit can be considered a `pro` but also a `con` depending on the POV

Reply:
To avoid confusion, as I was also wondering, I would avoid repeating the same phrase and instead go with something like:

```suggestion
1. Same as the pros, depending on the POV
```